packages/test-sync/src/androidAndroidTest/kotlin/io/realm/kotlin/test/mongodb/shared/ProgressListenerTests.kt

```diff
@@ -0,0 +1,249 @@
+            }
+
+            val flow = syncSession.progress(Direction.DOWNLOAD, ProgressMode.INDEFINITELY)
+            withTimeout(TIMEOUT) {
+                flow.first { it.isTransferComplete }
+            }
+        }.close()
+    }
+
+    @Test
+    fun worksAfterCancel() = runBlocking {
+        Realm.open(createSyncConfig(app.createUserAndLogIn())).apply {
+            writeSampleData(TEST_SIZE, timeout = TIMEOUT)
+        }.close()
+
+        Realm.open(createSyncConfig(app.createUserAndLogin())).apply {
+            // Setup a flow that we are just going to cancel
+            val flow = syncSession.progress(Direction.DOWNLOAD, ProgressMode.INDEFINITELY)
+            supervisorScope {
+                val task = async {
+                    flow.collect { /* no-op just keeping the flow active */ }
+                }
+                task.cancel()
+            }
+
+            // Verify that progress listeners still work
+            syncSession.progress(Direction.DOWNLOAD, ProgressMode.CURRENT_CHANGES).run {
+                withTimeout(TIMEOUT) {
+                    flow.first { it.isTransferComplete }
+                }
+            }
+        }.close()
+    }
+
+    @Test
+    fun addProgressListener_triggerImmediatelyWhenRegistered() = kotlinx.coroutines.runBlocking {
+        Realm.open(createSyncConfig(app.createUserAndLogIn())).apply {
+            withTimeout(10000) {
+                // Ensure that all data is already synced
+                assertTrue { syncSession.uploadAllLocalChanges() }
+                assertTrue { syncSession.downloadAllServerChanges() }
+                // Ensure that progress listeners are triggered at least one time even though there
+                // is no data
+                syncSession.progress(Direction.DOWNLOAD, ProgressMode.CURRENT_CHANGES).first()
+                syncSession.progress(Direction.UPLOAD, ProgressMode.CURRENT_CHANGES).first()
+                syncSession.progress(Direction.DOWNLOAD, ProgressMode.INDEFINITELY).first()
+                syncSession.progress(Direction.UPLOAD, ProgressMode.INDEFINITELY).first()
+            }
+        }.close()
+    }
+
+    private suspend fun Realm.writeSampleData(count: Int, idOffset: Int = 0, timeout: Duration? = null) {
+        write {
+            for (i in idOffset until count + idOffset) {
+                copyToRealm(ProgressTestObject().apply { _id = "Object $i" })
+            }
+        }
+        timeout?.let {
+            assertTrue { syncSession.uploadAllLocalChanges(timeout) }
+        }
+    }
+
+    // Operator that will return a flow that emits an incresing integer on each completion event
+    private fun Flow<Progress>.completionCounter(): Flow<Int> =
+        filter { it.isTransferComplete }
+            .distinctUntilChanged()
+            // Increment completed count if we are done transferring and the amount of bytes has
+            // increased
+            .scan(0UL to 0) { (bytes, completed), progress ->
```

Comment:
The initial value of `scan` is always emitted so that is why it is dropped. If there is ever an upstream `progress` it will emit a new event ... but should `Progress(0, 0)` then emit `completed=0` or `completed=1` 🤔 ... not sure it matters though as I have never seen `Progress(0, 0)` from core 😅